### ISSUE TITLE
`lambda` assignments in classes do not get their `self` parameter inferred

### ISSUE THREAD

**Issue by patrick-kidger** [microsoft/pyright#4504]
In the following snippet, the types of the `self` parameters are not correctly inferred:
```python
# information: Type of "self" is "Self@object"
# error: Cannot access member "foo" for type "object"
# Member "foo" is unknown (reportGeneralTypeIssues)
class A:
    foo = True
    __eq__ = lambda self, other: [reveal_type(self), self.foo][1]

# information: Type of "self" is "Unknown"
class B:
    foo = True
    __getitem__ = lambda self, item: [reveal_type(self), self.foo][1]
```
Expected behaviour is for:
- The `self` parameter to be inferred as `Self@A` and `Self@B` respectively.
- For `__eq__` and `__getitem__` to behave consistently. (Instead of inferring as `Self@object` an `Unknown`.)

---

For consistency I did also try the following. Each of them work, but their lack of brevity is the reason I was using a `lambda` in the first place.
```python
# information: Type of "self" is "Self@C"
class C:
    foo = True

    def __eq__(self, other):
        reveal_type(self)
        return self.foo

# information: Type of "self" is "Self@D"
class D:
    foo = True

    def __getitem__(self, other):
        reveal_type(self)
        return self.foo

from typing import Any, Callable, ClassVar, Self

# information: Type of "self" is "Self@E"
class E:
    foo = True
    __eq__: ClassVar[Callable[[Self, Any], bool]] = lambda self, other: [reveal_type(self), self.foo][1]
```

**Comment by erictraut** [comment]
Yes, this is by design. There isn't sufficient context for pyright to infer the type of lambda parameters if you are assigning them to an unannotated variable. You can use one of the recommended workarounds that you've listed above.

**Comment by patrick-kidger** [comment]
Why are `__eq__` and `__getitem__` treated inconsistently?

And, why is `__eq__ = lambda` any different to `def __eq__`? (I believe the same information will be available in both cases.)

**Comment by erictraut** [comment]
`__eq__` is defined in the base class (`object`) whereas `__getitem__` is not.

A `def` statement within a class has a specific semantic meaning. A `lambda` expression does not. The types of parameters in a lambda must be inferred using [bidirectional type inference](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#bidirectional-type-inference-expected-types), and that requires a declared type.

Pyright is working as designed here. This is not a bug. You'll need to use one of the other techniques if you want your code to type check without errors.

**Comment by patrick-kidger** [comment]
`__eq__` vs `__getitem__`: makes sense.

`def` vs `lambda`: I'm not aware of a spec anywhere that makes a distinction here. (Whether for static Python or just pyright.) In any case, if this distinction is desired behaviour for pyright, then I'll throw in some `# type: ignore`s to silence the errors.
